FreeRTOS-Plus/Demo/FreeRTOS-IoT-Libraries-LTS-Beta2/mqtt/mqtt_multitask/DemoTasks/MultitaskMQTTExample.c

```diff
@@ -0,0 +1,1494 @@
+/**
+ * @brief Notification bit used by subscriber task for unsubscribe operation.
+ */
+#define UNSUBSCRIBE_COMPLETE_BIT            ( 1U << 1 )
+
+/**
+ * @brief Maximum number of loop iterations to wait for a task notification.
+ */
+#define MAX_WAIT_ITERATIONS                 5
+
+/**
+ * @brief Topic filter used by the subscriber task.
+ */
+#define SUBSCRIBE_TOPIC_FILTER              "publish/+/filter"
+
+/**
+ * @brief Format string used by the publisher task for topic names.
+ */
+#define PUBLISH_TOPIC_FORMAT_STRING         "publish/%i/filter"
+
+/**
+ * @brief Format string used by the publisher task for payloads.
+ */
+#define PUBLISH_PAYLOAD_FORMAT              "Hello World! %d"
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief A type of command for interacting with the MQTT API.
+ */
+typedef enum CommandType
+{
+    PROCESSLOOP, /**< @brief Call MQTT_ProcessLoop(). */
+    PUBLISH,     /**< @brief Call MQTT_Publish(). */
+    SUBSCRIBE,   /**< @brief Call MQTT_Subscribe(). */
+    UNSUBSCRIBE, /**< @brief Call MQTT_Unsubscribe(). */
+    PING,        /**< @brief Call MQTT_Pint(). */
+    DISCONNECT,  /**< @brief Call MQTT_Disconnect(). */
+    CONNECT,     /**< @brief Placeholder command for reconnecting a broken connection. */
+    TERMINATE    /**< @brief Exit the command loop and stop processing commands. */
+} CommandType_t;
+
+/**
+ * @brief Struct containing context for a specific command.
+ *
+ * @note An instance of this struct and any variables it points to MUST stay
+ * in scope until the associated command is processed, and its callback called.
+ * The command callback will set the `complete` flag, and notify the calling task.
+ *
+ * @note This struct does not accept a parameter for number of subscriptions.
+ * Only a single subscription per SUBSCRIBE command is allowed in this demo.
+ * This is to simplify subscription management.
+ */
+typedef struct CommandContext
+{
+    MQTTPublishInfo_t * pPublishInfo;
+    MQTTSubscribeInfo_t * pSubscribeInfo;
+    size_t subscriptionCount;
+    MQTTStatus_t returnStatus;
+    bool complete;
+
+    /* The below fields are specific to this FreeRTOS implementation. */
+    TaskHandle_t taskToNotify;
+    uint32_t notificationBit;
+    QueueHandle_t pResponseQueue;
+} CommandContext_t;
+
+/**
+ * @brief Callback function called when a command completes.
+ */
+typedef void (* CommandCallback_t )( CommandContext_t * );
+
+/**
+ * @brief A command for interacting with the MQTT API.
+ */
+typedef struct Command
+{
+    CommandType_t commandType;
+    CommandContext_t * pContext;
+    CommandCallback_t callback;
+} Command_t;
+
+/**
+ * @brief Information for a pending MQTT ack packet expected by the demo.
+ */
+typedef struct ackInfo
+{
+    uint16_t packetId;
+    Command_t command;
+} AckInfo_t;
+
+/**
+ * @brief An element in the list of subscriptions maintained in the demo.
+ *
+ * @note This demo allows multiple tasks to subscribe to the same topic.
```

Comment:
I believe subscribing to the same topic by multiple tasks is supported by adding another element with same topic filter to the list. If so, can that be mentioned in the comments as well?

Reply:
Expanded on this comment.